src/provider/meta.ts

```diff
@@ -0,0 +1,63 @@
+import type { ExternalProvider, JsonRpcProvider, Web3Provider } from '@ethersproject/providers'
+import WalletConnectProvider from '@walletconnect/ethereum-provider'
+
+function isWeb3Provider(provider: JsonRpcProvider): provider is Web3Provider {
+  return 'provider' in provider
+}
+
+function isWalletConnectProvider(provider: ExternalProvider): provider is WalletConnectProvider {
+  return (provider as WalletConnectProvider).isWalletConnect
+}
+
+export enum WalletType {
+  WALLET_CONNECT = 'WalletConnect',
+  INJECTED = 'Injected',
+  UNKNOWN = 'Unknown',
+}
+
+export interface WalletMeta extends Partial<NonNullable<WalletConnectProvider['connector']['peerMeta']>> {
+  type: WalletType
+}
+
+function getWalletConnectMeta(provider: WalletConnectProvider): WalletMeta {
+  return {
+    type: WalletType.WALLET_CONNECT,
+    ...provider.connector.peerMeta,
+  }
+}
+
+function getInjectedMeta(provider: ExternalProvider & Record<string, unknown>): WalletMeta {
+  const properties = Object.getOwnPropertyNames(provider)
+  const names =
+    properties
+      .filter((name) => name.match(/^is.*$/) && (provider as Record<string, unknown>)[name])
```

Comment:
This should explicitly check for true to avoid including methods like isConnected.